File: DevEnvironment/DataStructures/src/test/java/datastructures/linkedlist/core/LinkedListTestData.java
package datastructures.linkedlist.core;

import base.core.ArrayLists;
import base.core.Triple;
import base.interfaces.ITriple;
import datastructures.linkedlist.interfaces.ILinkedListData;
import datastructures.linkedlist.interfaces.ILinkedListNode;
import datastructures.linkedlist.interfaces.ILinkedListTestData;
import java.util.ArrayList;
import java.util.List;

/**
 * The LinkedListTestData class implements data of tests of a linked list.
 */
public final class LinkedListTestData implements ILinkedListTestData {
    /**
     * The LinkedListTestData constructor.
     */
    public LinkedListTestData() {
    }

    /**
     * Gets the data of linked lists.
     */
    @Override
    public List<ILinkedListData<Integer>> getData() {
        List<ILinkedListData<Integer>> data = new ArrayList<>();

        data.add(this.getListData(0));
        data.add(this.getListData(50));
        data.add(this.getListData(100));

        return data;
    }

    /**
     * Gets the update data of a linked list.
     */
    @Override
    public List<ITriple<String, Integer, List<Integer>>> getUpdateData() {
        List<ITriple<String, Integer, List<Integer>>> data = new ArrayList<>();

        data.add(Triple.make("addToFront", 1, ArrayLists.make(1)));
        data.add(Triple.make("addToFront", 2, ArrayLists.make(2, 1)));
        data.add(Triple.make("addToFront", 3, ArrayLists.make(3, 2, 1)));
        data.add(Triple.make("removeFromFront", null, ArrayLists.make(2, 1)));
        data.add(Triple.make("removeFromFront", null, ArrayLists.make(1)));
        data.add(Triple.make("addToBack", 2, ArrayLists.make(1, 2)));
        data.add(Triple.make("addToBack", 3, ArrayLists.make(1, 2, 3)));
        data.add(Triple.make("addToBack", 4, ArrayLists.make(1, 2, 3, 4)));
        data.add(Triple.make("addToBack", 5, ArrayLists.make(1, 2, 3, 4, 5)));
        data.add(Triple.make("removeAfter", null, ArrayLists.make(1, 3, 4, 5)));
        data.add(Triple.make("removeAfter", null, ArrayLists.make(1, 4, 5)));
        data.add(Triple.make("removeAfter", null, ArrayLists.make(1, 5)));
        data.add(Triple.make("removeAfter", null, ArrayLists.make(1)));
        data.add(Triple.make("removeFromFront", null, new ArrayList<>()));

        return data;
    }

    /**
     * Gets the data of list.
     */
    private ILinkedListData<Integer> getListData(int size) {
        List<Integer> values = this.createListValues(size);
        List<ILinkedListNode<Integer>> nodes = this.createListNodes(values);

        return new LinkedListData<>(
            values,
            nodes);
    }

    /**
     * Creates values of a list.
     */
    private List<Integer> createListValues(int size) {
        List<Integer> result = new ArrayList<>();

        for (int i = 0; i < size; ++i) {
            int value = i + 1;
            result.add(value);
        }

        return result;
    }

    /**
     * Creates nodes of a list.
     */
    private <TValue extends Comparable<TValue>> List<ILinkedListNode<TValue>> createListNodes(List<TValue> values) {
        List<ILinkedListNode<TValue>> result = new ArrayList<>(values.size());

        for (TValue value : values) {
            result.add(LinkedListNode.make(value));
        }

        return result;
    }
}

File: DevEnvironmentCLion/DataStructuresTest/linkedlist/core/LinkedListTestData.cpp
#include "PreCompiled.h"

#include "LinkedListTestData.h"
#include "LinkedListData.h"

using namespace test::datastructures;
using namespace test::datastructures::linkedlist;

/**
 * The LinkedListTestData constructor.
 */
LinkedListTestData::LinkedListTestData()
{
}

/**
 * The LinkedListTestData destructor.
 */
LinkedListTestData::~LinkedListTestData()
{
}

/**
 * Gets the data of linked lists.
 */
std::vector<LinkedListDataSharedPtr<int>> LinkedListTestData::getData() const
{
    std::vector<LinkedListDataSharedPtr<int>> data;

    data.push_back(getListData(0));
    data.push_back(getListData(50));
    data.push_back(getListData(100));

    return data;
}

/**
 * Gets the updation data of a linked list.
 */
std::vector<std::tuple<std::string, int, std::vector<int>>> LinkedListTestData::getUpdationData()
{
    std::vector<std::tuple<std::string, int, std::vector<int>>> data;

    data.push_back(std::make_tuple<std::string, int, std::vector<int>>("addToFront", 1, {1}));
    data.push_back(std::make_tuple<std::string, int, std::vector<int>>("addToFront", 2, {2, 1}));
    data.push_back(std::make_tuple<std::string, int, std::vector<int>>("addToFront", 3, {3, 2, 1}));
    data.push_back(std::make_tuple<std::string, int, std::vector<int>>("removeFromFront", 0, {2, 1}));
    data.push_back(std::make_tuple<std::string, int, std::vector<int>>("removeFromFront", 0, {1}));
    data.push_back(std::make_tuple<std::string, int, std::vector<int>>("addToBack", 2, {1, 2}));
    data.push_back(std::make_tuple<std::string, int, std::vector<int>>("addToBack", 3, {1, 2, 3}));
    data.push_back(std::make_tuple<std::string, int, std::vector<int>>("addToBack", 4, {1, 2, 3, 4}));
    data.push_back(std::make_tuple<std::string, int, std::vector<int>>("addToBack", 5, {1, 2, 3, 4, 5}));
    data.push_back(std::make_tuple<std::string, int, std::vector<int>>("removeAfter", 0, {1, 3, 4, 5}));
    data.push_back(std::make_tuple<std::string, int, std::vector<int>>("removeAfter", 0, {1, 4, 5}));
    data.push_back(std::make_tuple<std::string, int, std::vector<int>>("removeAfter", 0, {1, 5}));
    data.push_back(std::make_tuple<std::string, int, std::vector<int>>("removeAfter", 0, {1}));
    data.push_back(std::make_tuple<std::string, int, std::vector<int>>("removeFromFront", 0, {}));

    return data;
}

/**
 * Gets the data of list.
 */
LinkedListDataSharedPtr<int> LinkedListTestData::getListData(size_t size) const
{
    std::vector<int> values = createListValues(size);
    std::vector<ILinkedListNodeSharedPtr<int>> nodes = createListNodes(values);

    LinkedListDataSharedPtr<int> data(new LinkedListData<int>(
        values,
        nodes));

    return data;
}

/**
 * Creates values of a list.
 */
std::vector<int> LinkedListTestData::createListValues(size_t size) const
{
    std::vector<int> result(size);

    for (size_t i = 0; i < size; ++i) {
        int value = i + 1;
        result[i] = value;
    }

    return result;
}

/**
 * Creates nodes of a list.
 */
template <typename T>
std::vector<ILinkedListNodeSharedPtr<T>> LinkedListTestData::createListNodes(const std::vector<T>& values) const
{
    std::vector<ILinkedListNodeSharedPtr<T>> result(values.size());

    for (size_t i = 0; i < values.size(); ++i)
    {
        T value = values[i];
        result[i] = LinkedListNode<T>::make(value);
    }

    return result;
}
